### PR TITLE
[WIP] Initial implementation of resource-control via systemd and dbus

### PR DESCRIPTION
**Pull request by nimrodshn**
## What this does?
Adds easy API for interacting with systemd via dbus (namely creating units), using this we can apply [resource control](https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html) via systemd.

Please note: I try to align with [runc's](https://github.com/opencontainers/runc/blob/6023d635d725a74c6eaa11ab7f3c870c073badd2/docs/systemd.md#systemd-cgroup-driver) approach in this line of work to walk in the steps of giants.

TODO: 

- [ ] Implement the "controllers" abstraction for each resource supported by `systemd` properties. (Traits and respective implementation).
- [ ] Finish implementing `start_transient_unit_for_container` which eventually applies the properties to the newly created unit associated with the container (the last `for` loop doesn't compile in the current vesion :cry: ). The bindings generated to the dbus interfaces are used using the following command: ```> dbus-codegen-rust -s -g -m None -d org.freedesktop.systemd1 -p /org/freedesktop/systemd1``` (See https://github.com/diwic/dbus-rs)

Related issue: #24 

cc: @utam0k @Furisto

### CODE REVIEW

File: src/dbus/client.rs [hunk 1]
```diff
@@ -15,19 +18,90 @@ impl Client {
...
+        } else {
+            properties.push(PropSlice::new(slice).into());
+        }
+
+        proxy.start_transient_unit(unit_name, "replace", properties, vec![])?;
         Ok(())
     }
 }
+
+type Property = (&'static str, Variant<Box<dyn RefArg>>);
```
Comment (nimrodshn) [inline src/dbus/client.rs:56]: I'm considering using a procedural macro to derive all these helper structs and their impls.

Comment (nimrodshn) [reply src/dbus/client.rs:56]: Or I can keep it simple and just write the tuples directly.

**Comment by utam0k** [comment]
@nimrodshn Is there anything I can do to help?

**Comment by nimrodshn** [comment]
@utam0k Sorry I was busy with other projects :cry: I will try and get some work on this soon..

**Comment by utam0k** [comment]
@nimrodshn I understand. You can do it at your own pace. I was a little worried because you've been contributing to youki quite a bit.

**Comment by nimrodshn** [comment]
@utam0k  Sorry I was busy with life :sweat_smile: . I will try to advance this asap.

@Furisto suggested I make the resource control part modular (similar to how we provide `Controller` impls for each controller type, e.g. memory, io etc.) How do you envision that API work here? Should we pass a list of controllers to `start_transient_unit_for_container` in addition to `LinuxResources`? As These are provided statically in each cgroups manager (For example [here](https://github.com/containers/youki/blob/main/src/cgroups/v2/manager.rs#L28-L36).) - I think that might be a simple solution that way this dbus client could be used both for `v1` and `v2` systemd cgroup managers.

**Comment by Furisto** [comment]
@nimrodshn My idea was that the dbus client should not know anything about resources or controllers. The responsibility of the dbus client is to communicate with systemd and nothing else. The responsibility of the controllers is to translate the resource restrictions described in the spec to systemd properties. 
That means they get passed in a `Resource` and return back properties. We collect these properties from the controllers (and maybe additional ones that are created by the cgroup manager directly like for pids) and hand them over do the dbus client. Should systemd support further resources in the future we can just add another controller without having to touch the surrounding code. 

You do not need to implement it exactly like this though, if you think you have a better idea, go for it.

**Comment by Furisto** [comment]
@nimrodshn Can I support you in any way?

**Comment by nimrodshn** [comment]
@Furisto No I'm good thanks! I just came back to this :smile:

**Comment by utam0k** [comment]
@nimrodshn Hi! Can I ask you to write out the remaining work?

**Comment by nimrodshn** [comment]
@utam0k  Wrote the work to be done in the PRs description :arrow_up:

**Comment by utam0k** [comment]
@nimrodshn Thanks! Let me know if you are busy or need any help. This isn't going to be included in the first release of youki, so taking it slow is fine. Please don't feel pressured. We need your help, I'll be waiting for you to come back ;)

**Comment by nimrodshn** [comment]
Thanks @utam0k currently finding it hard to work on open source :crying_cat_face: as I'm in between jobs so taking it one step at a time :smile_cat:

**Comment by utam0k** [comment]
> Thanks @utam0k currently finding it hard to work on open source  as I'm in between jobs so taking it one step at a time

Of course, there are times like that. No problem. I'm waiting for this PR to be open, so let me know when you have more time. If you have plans, please let me know.

**Comment by utam0k** [comment]
We'll continue here. @nimrodshn Thanks a lot.
#451

**Pull request closed without merging**